microcosm_pubsub/tests/test_publish_decorator.py

```diff
@@ -0,0 +1,198 @@
+        assert_that(loads(self.graph.sns_producer.sns_client.publish.call_args[1]["Message"]), is_(equal_to({
+            "mediaType": "application/vnd.globality.pubsub._.created.company",
+            "uri": "http://localhost/api/v1/company/ID",
+            "opaqueData": {},
+        })))
+
+    def test_set_media_type(self):
+        decorator = publish(media_type=changed("something"))
+        with self.graph.app.test_request_context():
+            create_decorated_controller(self.graph, decorator).retrieve()
+        assert_that(loads(self.graph.sns_producer.sns_client.publish.call_args[1]["Message"]), is_(equal_to({
+            "mediaType": "application/vnd.globality.pubsub._.changed.something",
+            "uri": "http://localhost/api/v1/company/ID",
+            "opaqueData": {},
+        })))
+
+    def test_set_uri_string_args(self):
+        decorator = publish(uri_string_args={
+            "name": lambda model: model.name,
+            "limit": lambda model: 1,
+            "company_id": lambda model: model.id,
+        })
+        with self.graph.app.test_request_context():
+            create_decorated_controller(self.graph, decorator).retrieve()
+        assert_that(loads(self.graph.sns_producer.sns_client.publish.call_args[1]["Message"]), is_(equal_to({
+            "mediaType": "application/vnd.globality.pubsub._.created.company",
+            "uri": "http://localhost/api/v1/company/ID?name=Name&limit=1",
+            "opaqueData": {},
+        })))
+
+    def test_no_uri(self):
+        decorator = publish(default_uri=False)
+        create_decorated_controller(self.graph, decorator).retrieve()
+        assert_that(loads(self.graph.sns_producer.sns_client.publish.call_args[1]["Message"]), is_(equal_to({
+            "mediaType": "application/vnd.globality.pubsub._.created.company",
+            "opaqueData": {},
+        })))
+
+    def test_custom_uri(self):
+        decorator = publish(default_uri=False, message_params={"uri": lambda model: f"http://example.com/{model.id}"})
+        create_decorated_controller(self.graph, decorator).retrieve()
+        assert_that(loads(self.graph.sns_producer.sns_client.publish.call_args[1]["Message"]), is_(equal_to({
+            "mediaType": "application/vnd.globality.pubsub._.created.company",
+            "uri": "http://example.com/ID",
+            "opaqueData": {},
+        })))
+
+    def test_set_message_params(self):
+        decorator = publish(media_type=DuckTypeSchema.MEDIA_TYPE, message_params={"quack": lambda model: model.name})
+        with self.graph.app.test_request_context():
+            create_decorated_controller(self.graph, decorator).retrieve()
+        assert_that(loads(self.graph.sns_producer.sns_client.publish.call_args[1]["Message"]), is_(equal_to({
+            "quack": "Name",
+        })))
+
+    def test_set_operation(self):
+        decorator = publish(operation=Operation.Search)
```

Comment:
I think it might be smart enough but how does this handle one of the RetrieveFor/SearchFor operations?